test/vm_tests/builder/declaration_parsing_test.dart

```diff
@@ -91,430 +112,1414 @@ main() {
-              expect(declarations.state.meta.keyNamespace, 'baz');
-              expect(declarations.component2.meta.isWrapper, isTrue);
+                componentDeclarations = declarations.whereType<ClassComponentDeclaration>();
+              });
+
+              test('detects the correct number of components', () {
+                expect(componentDeclarations.length, 2,
+                    reason: 'There should be two different components.');
+              });
+
+              test('with names that match expectations', () {
+                final firstComponent = componentDeclarations.firstWhereNameEquals('FirstFoo');
+                final secondComponent = componentDeclarations.firstWhereNameEquals('SecondFoo');
+                expect(firstComponent, isNotNull);
+                expect(secondComponent, isNotNull);
+              });
+
+              test('that are the correct version', () {
+                final versionForFirstMixin =
+                resolveVersion(componentDeclarations.firstWhereNameEquals('FirstFoo').members);
+                final versionForSecondMixin =
+                resolveVersion(componentDeclarations.firstWhereNameEquals('SecondFoo').members);
+
+                expect(versionForFirstMixin.version, Version.v4_mixinBased);
+                expect(versionForSecondMixin.version, Version.v4_mixinBased);
+              });
             });
 
-            test('a props mixin', () {
-              setUpAndParse('''
-                @PropsMixin(keyNamespace: "bar")
-                class FooPropsMixin {}
+            group('and there is a mix of abstract and concrete', () {
+              Iterable<LegacyClassComponentDeclaration> componentDeclarations;
+
+              setUp(() {
+                setUpAndParse('''
+                ${OverReactSrc.abstractState(needsComponent: false, baseName: 'StateFoo').source}
+                
+                ${OverReactSrc.abstractProps(needsComponent: false, baseName: 'PropsFoo').source}
+  
+                ${getBoilerplateString(version: BoilerplateVersions.v2, componentBaseName: 'SecondFoo')}
+                
+                ${getBoilerplateString(version: BoilerplateVersions.v3, componentBaseName: 'ThirdFoo')}
               ''');
-              expect(declarations.propsMixins.single.meta.keyNamespace, 'bar');
+
+                componentDeclarations = declarations.whereType<LegacyClassComponentDeclaration>();
+              });
+
+              test('detects the correct number of components', () {
+                expect(componentDeclarations.length, 2,
+                    reason: 'There should be two different components.');
+              });
+
+              test('with names that match expectations', () {
+                final propsComponentDeclaration =
+                componentDeclarations.firstWhereNameEquals('PropsFoo');
+                final stateComponentDeclaration =
+                componentDeclarations.firstWhereNameEquals('StateFoo');
```

Comment:
#superFormattingNit these could use the two extra tabs (can't remember the term for the tabbing expected when an expression continues onto a second line).
```suggestion
                    componentDeclarations.firstWhereNameEquals('PropsFoo');
                final stateComponentDeclaration =
                    componentDeclarations.firstWhereNameEquals('StateFoo');
```